metricbeat/module/system/diskio/diskstat_linux.go

```diff
@@ -34,7 +34,81 @@ func Get_CLK_TCK() uint32 {
 
 // IOCounters should map functionality to disk package for linux os.
 func IOCounters(names ...string) (map[string]disk.IOCountersStat, error) {
-	return disk.IOCounters(names...)
+	stats, err := disk.IOCounters(names...)
+	if err != nil {
+		return nil, err
+	}
+	//Process `stats`, as `names` might be empty
+	topCounters := separateTopLevelCounters(stats)
+	stats["summary"] = summaryIOCounter(topCounters)
+	return stats, nil
+}
+
+func summaryIOCounter(stats map[string]disk.IOCountersStat) disk.IOCountersStat {
+	sum := disk.IOCountersStat{}
+	sum.Name = "summary"
+
+	for _, counter := range stats {
+		sum.ReadCount += counter.ReadCount
+		sum.ReadBytes += counter.ReadBytes
+		sum.ReadTime += counter.ReadTime
+		sum.WriteCount += counter.WriteCount
+		sum.WriteBytes += counter.WriteBytes
+		sum.WriteTime += counter.WriteTime
+		sum.IoTime += counter.IoTime
+		sum.MergedReadCount += counter.MergedReadCount
+		sum.MergedWriteCount += counter.MergedWriteCount
+		sum.WeightedIO += counter.WeightedIO
+	}
+
+	return sum
+}
+
+// separateTopLevelCounters will return a map of top level counters to ,
+// and it removes only aggregate counters from the passed map `stats`.
+// for example, it adds `hda` to a new map, and will not remove `hda`
+// if there is no sub counters such as `hda1`
+func separateTopLevelCounters(stats map[string]disk.IOCountersStat) map[string]disk.IOCountersStat {
+	separated := map[string]disk.IOCountersStat{}
+	for name := range stats {
+		i := len(name)
+
+		//Skip the index to the first digit of the tailing number
+		for name[i-1] <= '9' && name[i-1] >= '0' {
+			i--
+		}
+
+		if i == len(name) {
+			separated[name] = stats[name]
+			continue
+		}
+
+		// found is true if the high level disk name of the given one is found.
+		// This handles the situation that `m.includeDevices` contains sub level names only
+		found := false
+
+		//for SATA and IDE disks, the names are like `sda`, `sda1`, `sda2`
+		//for nvme disks, the names would be like `nvme0n1` and `nvme0n1p1`
+		for j := i; j > i-2 && !found; j-- {
+			if _, ok := separated[name[:j]]; ok {
+				found = true
+			}
+
+			if _, ok := stats[name[:j]]; ok {
```

Comment:
Yes, you are right about the logic and thank you for the advice on refactoring. That looks much better indeed. :-)

Reply:
Would you mind doing the refactoring part after, as your codes seem to work perfectly and that's your work?

Reply:
Well, no it needs to be cleaned up before being checked in the first time, sorry :-) It will likely still need more work... like I said, we will need to think more about the "right" way to classify and report parent-child connections in the device names (to make sure we don't surprise people with changed behavior, that we're reporting it in a useful way, etc.)

The first-pass review was to get it in a clean initial state for that conversation. I'm happy to help you with the process if you'd like to proceed, but it's also ok to step back if this is more than you bargained for... either way, thank you for looking at the issue!

Reply:
@faec  You are right, it might still need more work. 
I'd love to proceed. Thank you for your reviews, advice and help. :smiley: 

I'd like to document more specifics first to see if it makes things clear.